File: bench/python/bench_kernel.py
#!/usr/bin/env python3
"""Kernel / GP benchmarks (B2) — Python side.

Run with::

    OPENBLAS_NUM_THREADS=1 OMP_NUM_THREADS=1 \
        bench/venv/bin/python bench/python/bench_kernel.py
"""

from __future__ import annotations

import csv
import time
from dataclasses import dataclass
from pathlib import Path

import numpy as np
import pandas as pd
from sklearn.kernel_ridge import KernelRidge
from sklearn.kernel_approximation import RBFSampler
from sklearn.linear_model import Ridge
from sklearn.metrics.pairwise import rbf_kernel
from sklearn.gaussian_process import GaussianProcessRegressor
from sklearn.gaussian_process.kernels import RBF, ConstantKernel as C, WhiteKernel


REPO = Path(__file__).resolve().parents[2]
DATA = REPO / "bench" / "data"
OUT = REPO / "bench" / "results" / "python"
OUT.mkdir(parents=True, exist_ok=True)

H = 1.0
LAM = 1e-3
GAMMA = 1.0 / (2 * H * H)  # rbf_kernel uses gamma = 1/(2h²)


@dataclass
class Row:
    name: str
    time_ms: float
    acc_main: float
    acc_aux: float
    extra: str


def median_time(fn, n_iter: int) -> tuple[float, object]:
    times, last = [], None
    for _ in range(n_iter):
        t0 = time.perf_counter()
        last = fn()
        t1 = time.perf_counter()
        times.append(1000.0 * (t1 - t0))
    times.sort()
    return times[len(times) // 2], last


def load_xy(path: Path):
    df = pd.read_csv(path)
    return df.iloc[:, :-1].to_numpy(), df.iloc[:, -1].to_numpy()


def r2(y, yhat):
    mu = y.mean()
    sst = float(((y - mu) ** 2).sum())
    sse = float(((y - yhat) ** 2).sum())
    return 0.0 if sst == 0 else 1 - sse / sst


def rmse(y, yhat):
    return float(np.sqrt(((y - yhat) ** 2).mean()))


# ---------------------------------------------------------------------------

def bench_gram(path: Path, name: str) -> Row:
    x, _ = load_xy(path)

    def run():
        return rbf_kernel(x, x, gamma=GAMMA)

    ms, _ = median_time(run, 5)
    return Row(name, ms, 0.0, 0.0, f"sklearn.rbf_kernel n={x.shape[0]}")


def bench_kr(path: Path, name: str) -> Row:
    x, y = load_xy(path)

    def run():
        return KernelRidge(alpha=LAM, kernel="rbf", gamma=GAMMA).fit(x, y)

    ms, m = median_time(run, 3)
    yhat = m.predict(x)
    return Row(name, ms, r2(y, yhat), rmse(y, yhat),
               "sklearn.KernelRidge RBF")


def bench_nw(path: Path, name: str) -> Row:
    x, y = load_xy(path)

    def run():
        K = rbf_kernel(x, x, gamma=GAMMA)
        rowsum = K.sum(axis=1)
        rowsum[rowsum == 0] = 1.0
        return (K @ y) / rowsum

    ms, yhat = median_time(run, 3)
    return Row(name, ms, r2(y, yhat), rmse(y, yhat),
               "manual NW via rbf_kernel")


def bench_rff(path: Path, name: str, d: int) -> Row:
    x, y = load_xy(path)
    sampler = RBFSampler(gamma=GAMMA, n_components=d, random_state=0).fit(x)
    phi = sampler.transform(x)

    def run():
        return Ridge(alpha=LAM, fit_intercept=False).fit(phi, y)

    ms, m = median_time(run, 3)
    yhat = m.predict(phi)
    return Row(name, ms, r2(y, yhat), rmse(y, yhat),
               f"sklearn.RBFSampler+Ridge D={d}")


def bench_gp_fit(path: Path, name: str) -> Row:
    x, y = load_xy(path)
    kernel = C(1.0, constant_value_bounds="fixed") \
             * RBF(length_scale=H, length_scale_bounds="fixed") \
             + WhiteKernel(noise_level=0.05, noise_level_bounds="fixed")

    def run():
        gpr = GaussianProcessRegressor(kernel=kernel, optimizer=None,
                                       normalize_y=False)
        gpr.fit(x, y)
        return gpr

    ms, gpr = median_time(run, 3)
    yhat, std = gpr.predict(x, return_std=True)
    return Row(name, ms, r2(y, yhat), rmse(y, yhat),
               "sklearn.GaussianProcessRegressor RBF (HP fixed)")


def bench_gp_opt(path: Path, name: str) -> Row:
    x, y = load_xy(path)
    kernel = C(1.0) * RBF(length_scale=0.5) + WhiteKernel(noise_level=0.05)

    def run():
        gpr = GaussianProcessRegressor(kernel=kernel, normalize_y=False,
                                       n_restarts_optimizer=0)
        gpr.fit(x, y)
        return gpr

    ms, gpr = median_time(run, 2)
    yhat = gpr.predict(x)
    # length scale extracted from optimized kernel
    try:
        ell = float(gpr.kernel_.k1.k2.length_scale)
    except Exception:
        ell = float("nan")
    return Row(name, ms, r2(y, yhat), ell,
               "sklearn GP fit + optimize (L-BFGS-B)")


# ---------------------------------------------------------------------------

def main() -> int:
    rows: list[Row] = [
        bench_gram(DATA / "kernel_n500_p5.csv",  "GramMV_n500_p5"),
        bench_gram(DATA / "kernel_n1000_p5.csv", "GramMV_n1000_p5"),
        bench_gram(DATA / "kernel_n2000_p5.csv", "GramMV_n2000_p5"),
        bench_kr(DATA / "kernel_n500_p5.csv",  "KR_n500_p5"),
        bench_kr(DATA / "kernel_n1000_p5.csv", "KR_n1000_p5"),
        bench_kr(DATA / "kernel_n2000_p5.csv", "KR_n2000_p5"),
        bench_nw(DATA / "kernel_n1000_p5.csv", "NW_n1000_p5"),
        bench_rff(DATA / "kernel_n1000_p5.csv", "RFF_n1000_D256_p5", 256),
        bench_rff(DATA / "kernel_n2000_p5.csv", "RFF_n2000_D256_p5", 256),
        bench_gp_fit(DATA / "kernel_n500_p5.csv",  "GP_fit_n500_p5"),
        bench_gp_fit(DATA / "kernel_n1000_p5.csv", "GP_fit_n1000_p5"),
        bench_gp_opt(DATA / "kernel_n500_p5.csv",  "GP_opt_n500_p5"),
        # GPRobust: no direct sklearn equivalent — skipped.
    ]
    out = OUT / "kernel.csv"
    with open(out, "w", newline="") as fh:
        w = csv.writer(fh)
        w.writerow(["system", "suite", "name",
                    "time_ms", "acc_main", "acc_aux", "extra"])
        for r in rows:
            w.writerow(["python", "kernel", r.name,
                        f"{r.time_ms:.6g}", f"{r.acc_main:.6g}",
                        f"{r.acc_aux:.6g}", r.extra])
    print(f"wrote {len(rows)} rows -> {out}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: bench/haskell/BenchKernel.hs
{-# LANGUAGE OverloadedStrings #-}
{-# OPTIONS_GHC -fno-full-laziness -fno-cse #-}
-- | Kernel / GP benchmarks (B2).

module Main where

import qualified Numeric.LinearAlgebra   as LA
import qualified System.Random.MWC       as MWC
import qualified Data.Vector             as V

import qualified Hanalyze.Model.Kernel            as Kn
import qualified Hanalyze.Model.GP                as GP
import qualified Hanalyze.Model.RFF               as RFF
import qualified Hanalyze.Model.GPRobust          as GPR

import           BenchUtil

-- ---------------------------------------------------------------------------

main :: IO ()
main = do
  rows <- mconcat <$> sequence
    [ benchGram     "bench/data/kernel_n500_p5.csv"  "GramMV_n500_p5"
    , benchGram     "bench/data/kernel_n1000_p5.csv" "GramMV_n1000_p5"
    , benchGram     "bench/data/kernel_n2000_p5.csv" "GramMV_n2000_p5"
    , benchGram     "bench/data/kernel_n4000_p5.csv" "GramMV_n4000_p5"
    , benchKR       "bench/data/kernel_n500_p5.csv"  "KR_n500_p5"
    , benchKR       "bench/data/kernel_n1000_p5.csv" "KR_n1000_p5"
    , benchKR       "bench/data/kernel_n2000_p5.csv" "KR_n2000_p5"
    , benchKR       "bench/data/kernel_n4000_p5.csv" "KR_n4000_p5"
    , benchNW       "bench/data/kernel_n1000_p5.csv" "NW_n1000_p5"
    , benchRFF      "bench/data/kernel_n1000_p5.csv" "RFF_n1000_D256_p5"  256
    , benchRFF      "bench/data/kernel_n2000_p5.csv" "RFF_n2000_D256_p5"  256
    , benchGPFit    "bench/data/kernel_n500_p5.csv"  "GP_fit_n500_p5"
    , benchGPFit    "bench/data/kernel_n1000_p5.csv" "GP_fit_n1000_p5"
    , benchGPFit    "bench/data/kernel_n2000_p5.csv" "GP_fit_n2000_p5"
    , benchGPOpt    "bench/data/kernel_n500_p5.csv"  "GP_opt_n500_p5"
    , benchGPRobust "bench/data/kernel_n500_p5.csv"  "GPRobust_n500_p5"
    ]
  writeRows "bench/results/haskell/kernel.csv" rows
  putStrLn $ "wrote " ++ show (length rows)
          ++ " rows → bench/results/haskell/kernel.csv"

-- ---------------------------------------------------------------------------
-- 共通設定: Gaussian RBF, h = 1.0, λ = 1e-3
h0 :: Double
h0 = 1.0

lam0 :: Double
lam0 = 1e-3

-- ---------------------------------------------------------------------------
-- Gram matrix (BLAS pairwise dist + cmap)
-- ---------------------------------------------------------------------------

{-# NOINLINE gramPhantom #-}
gramPhantom :: Int -> Kn.Kernel -> Double
            -> LA.Matrix Double -> LA.Matrix Double
gramPhantom _ k h x = Kn.gramMatrixMV k h x

benchGram :: FilePath -> String -> IO [BenchRow]
benchGram path name = do
  (x, _) <- readCsvXY path
  (ms, g) <- timeitTastyIO LA.sumElements
               (\i -> return $! gramPhantom i Kn.Gaussian h0 x)
  return [ BenchRow "haskell" "kernel" name ms 0 0
            ("gramMatrixMV BLAS, n=" ++ show (LA.rows g)) ]

-- ---------------------------------------------------------------------------
-- Kernel Ridge fit (multi-input)
-- ---------------------------------------------------------------------------

{-# NOINLINE krPhantom #-}
krPhantom :: Int -> LA.Matrix Double -> LA.Matrix Double -> Kn.KernelRidgeFitMV
krPhantom _ x ym = Kn.kernelRidgeMV Kn.Gaussian h0 lam0 x ym

benchKR :: FilePath -> String -> IO [BenchRow]
benchKR path name = do
  (x, y) <- readCsvXY path
  let yMat = LA.asColumn y
  (ms, fit) <- timeitTastyIO (\f -> LA.sumElements (Kn.krmvAlpha f))
                 (\i -> return $! krPhantom i x yMat)
  let yhat = LA.flatten (Kn.fittedKernelRidgeMV fit LA.¿ [0])
      r2v  = computeR2 y yhat
  return [ BenchRow "haskell" "kernel" name ms r2v
                     (sqrt (LA.sumElements ((y - yhat) ** 2)
                            / fromIntegral (LA.size y)))
                     ("kernelRidgeMV Gaussian h=1 λ=1e-3") ]

-- ---------------------------------------------------------------------------
-- Nadaraya-Watson
-- ---------------------------------------------------------------------------

{-# NOINLINE nwPhantom #-}
nwPhantom :: Int -> LA.Matrix Double -> LA.Matrix Double -> LA.Matrix Double
nwPhantom _ x ym = Kn.nwRegressionMV Kn.Gaussian h0 x ym x

benchNW :: FilePath -> String -> IO [BenchRow]
benchNW path name = do
  (x, y) <- readCsvXY path
  let yMat = LA.asColumn y
  (ms, yhatMat) <- timeitTastyIO LA.sumElements
                     (\i -> return $! nwPhantom i x yMat)
  let yhat = LA.flatten (yhatMat LA.¿ [0])
      r2v  = computeR2 y yhat
  return [ BenchRow "haskell" "kernel" name ms r2v
                     (sqrt (LA.sumElements ((y - yhat) ** 2)
                            / fromIntegral (LA.size y)))
                     "nwRegressionMV Gaussian h=1" ]

-- ---------------------------------------------------------------------------
-- RFF Ridge (multivariate input)
-- ---------------------------------------------------------------------------

{-# NOINLINE rffPhantom #-}
rffPhantom :: Int -> RFF.RFFFeaturesMV -> LA.Matrix Double
           -> LA.Matrix Double -> RFF.RFFRidgeFitMVMO
rffPhantom _ feats x ym = RFF.rffRidgeMVMulti feats x ym lam0

benchRFF :: FilePath -> String -> Int -> IO [BenchRow]
benchRFF path name d = do
  (x, y) <- readCsvXY path
  let ym = LA.asColumn y
      p  = LA.cols x
  gen <- MWC.createSystemRandom
  feats <- RFF.sampleRFFRBFMV p d 1.0 1.0 gen
  (ms, _) <- timeitTastyIO (\f -> LA.sumElements (RFF.rffrmvmWeights f))
                (\i -> return $! rffPhantom i feats x ym)
  let yhatMat = RFF.predictRFFRidgeMVMulti
                  (rffPhantom 0 feats x ym) x
      yhat = LA.flatten (yhatMat LA.¿ [0])
      r2v  = computeR2 y yhat
  return [ BenchRow "haskell" "kernel" name ms r2v
                     (sqrt (LA.sumElements ((y - yhat) ** 2)
                            / fromIntegral (LA.size y)))
                     ("RFFFeaturesMV D=" ++ show d) ]

-- ---------------------------------------------------------------------------
-- GP fit (HP fixed)
-- ---------------------------------------------------------------------------

{-# NOINLINE gpFitPhantom #-}
gpFitPhantom :: Int -> GP.GPModel
             -> LA.Matrix Double -> LA.Vector Double -> LA.Matrix Double
             -> GP.GPResultMV
gpFitPhantom _ mdl x y t = GP.fitGPMV mdl x y t

benchGPFit :: FilePath -> String -> IO [BenchRow]
benchGPFit path name = do
  (x, y) <- readCsvXY path
  let mdl = GP.GPModel GP.RBF (GP.GPParams 1.0 1.0 0.05 1.0 Nothing)
  (ms, res) <- timeitTastyIO (\r -> LA.sumElements (GP.gpmvMean r)
                                + LA.sumElements (GP.gpmvVar r))
                 (\i -> return $! gpFitPhantom i mdl x y x)
  let yhat = GP.gpmvMean res
      r2v  = computeR2 y yhat
  return [ BenchRow "haskell" "kernel" name ms r2v
                     (sqrt (LA.sumElements ((y - yhat) ** 2)
                            / fromIntegral (LA.size y)))
                     "fitGPMV RBF (HP fixed)" ]

-- ---------------------------------------------------------------------------
-- GP HP optimization (L-BFGS over log marginal likelihood)
-- ---------------------------------------------------------------------------

{-# NOINLINE gpOptPhantom #-}
gpOptPhantom :: Int -> LA.Matrix Double -> LA.Vector Double -> GP.GPParams
gpOptPhantom _ x y = GP.optimizeGPMV GP.RBF x y
                       (GP.GPParams 0.5 1.0 0.05 1.0 Nothing)

benchGPOpt :: FilePath -> String -> IO [BenchRow]
benchGPOpt path name = do
  (x, y) <- readCsvXY path
  (ms, p) <- timeitTastyIO (\pr -> GP.gpLengthScale pr + GP.gpSignalVar pr
                              + GP.gpNoiseVar pr)
               (\i -> return $! gpOptPhantom i x y)
  let mdl = GP.GPModel GP.RBF p
      res = GP.fitGPMV mdl x y x
      yhat = GP.gpmvMean res
      r2v  = computeR2 y yhat
  return [ BenchRow "haskell" "kernel" name ms r2v (GP.gpLengthScale p)
                     "optimizeGPMV (L-BFGS / log marginal likelihood)" ]

-- ---------------------------------------------------------------------------
-- GPRobust IRLS (Student-t)
-- ---------------------------------------------------------------------------

{-# NOINLINE gprPhantom #-}
gprPhantom :: Int -> LA.Matrix Double -> LA.Vector Double
           -> GPR.RobustGPFitMV
gprPhantom _ x y = GPR.fitGPRobustMV GP.RBF
                      (GP.GPParams 1.0 1.0 0.05 1.0 Nothing)
                      (GPR.RStudentT 4.0 0.1)
                      x y

benchGPRobust :: FilePath -> String -> IO [BenchRow]
benchGPRobust path name = do
  (x, y) <- readCsvXY path
  (ms, fit) <- timeitTastyIO (\f -> LA.sumElements (GPR.rgpmvAlpha f))
                 (\i -> return $! gprPhantom i x y)
  let (mu, _) = GPR.predictGPRobustMV fit x
      r2v    = computeR2 y mu
  return [ BenchRow "haskell" "kernel" name ms r2v (fromIntegral (GPR.rgpmvIters fit))
                     "fitGPRobustMV StudentT(4, 0.1)" ]

-- ---------------------------------------------------------------------------

computeR2 :: LA.Vector Double -> LA.Vector Double -> Double
computeR2 y yhat =
  let mu  = LA.sumElements y / fromIntegral (LA.size y)
      sst = LA.sumElements ((y - LA.konst mu (LA.size y)) ** 2)
      sse = LA.sumElements ((y - yhat) ** 2)
  in if sst == 0 then 0 else 1 - sse / sst
